### PR TITLE
gh-89532: Remove LibreSSL workarounds

### LINKED ISSUE #89532: Remove LibreSSL workarounds
BPO | [45369](https://bugs.python.org/issue45369)
--- | :---
Nosy | @tiran, @ramikg
PRs | <li>python/cpython#28728</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/tiran'
closed_at = None
created_at = <Date 2021-10-05.04:56:22.211>
labels = ['expert-SSL', 'type-feature', '3.10', '3.11']
title = 'Remove LibreSSL workarounds'
updated_at = <Date 2021-10-05.08:45:15.966>
user = 'https://github.com/ramikg'
```

bugs.python.org fields:
```python
activity = <Date 2021-10-05.08:45:15.966>
actor = 'christian.heimes'
assignee = 'christian.heimes'
closed = False
closed_date = None
closer = None
components = ['SSL']
creation = <Date 2021-10-05.04:56:22.211>
creator = 'ramikg'
dependencies = []
files = []
hgrepos = []
issue_num = 45369
keywords = ['patch']
message_count = 1.0
messages = ['403193']
nosy_count = 2.0
nosy_names = ['christian.heimes', 'ramikg']
pr_nums = ['28728']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue45369'
versions = ['Python 3.10', 'Python 3.11']
```

</p></details>

### PR DESCRIPTION
**Pull request by ramikg**
Python 3.10 drops support for LibreSSL (as per [PEP 644](https://www.python.org/dev/peps/pep-0644/)), but there are still a few workarounds & mentions of LibreSSL in the source code.

The related PR cleans the code of any LibreSSL workarounds or mentions.

https://bugs.python.org/issue45369

* Issue: gh-89532

Fixes #89532

### CODE REVIEW

File: Lib/test/test_ssl.py [hunk 1]
```diff
@@ -539,10 +539,9 @@ def test_openssl_version(self):
         self.assertGreaterEqual(status, 0)
         self.assertLessEqual(status, 15)
 
-        libressl_ver = f"LibreSSL {major:d}"
         openssl_ver = f"OpenSSL {major:d}.{minor:d}.{fix:d}"
         self.assertTrue(
-            s.startswith((openssl_ver, libressl_ver)),
```
Comment (tiran) [inline Lib/test/test_ssl.py]: Please leave this check.

File: Misc/NEWS.d/next/Core and Builtins/2021-10-05-05-00-16.bpo-45369.tluk_X.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Remove LibreSSL support as per PEP 644.
```
Comment (tiran) [inline Misc/NEWS.d/next/Core and Builtins/2021-10-05-05-00-16.bpo-45369.tluk_X.rst]: Your wording is a bit strong. Python will support LibreSSL again as soon as it becomes API compatible with OpenSSL 1.1.1.

```suggestion
Remove LibreSSL workarounds as per :pep`644`.
```

File: Tools/ssl/multissltests.py [hunk 3]
```diff
@@ -1,12 +1,12 @@
 #!./python
-"""Run Python tests against multiple installations of OpenSSL and LibreSSL
```
Comment (tiran) [inline Tools/ssl/multissltests.py]: Please leave this test helper unmodified. The LibreSSL code will be useful as soon as LibreSSL becomes API compatible with OpenSSL 1.1.1.

**Review by tiran — CHANGES_REQUESTED** [review]
Thanks for your PR.

Python 3.10 didn't actively drop support for LibreSSL. For a lack of better wording, we *passively* removed compatibility with forks that don't provide OpenSSL 1.1.1 API. I still like to keep some code so we can re-add LibreSSL support later.

**Review by tiran — COMMENTED** [review]
Thanks for the quick fix!

**Comment by dhewg** [comment]
> For a lack of better wording, we _passively_ removed compatibility with forks that don't provide OpenSSL 1.1.1 API. I still like to keep some code so we can re-add LibreSSL support later.

Some time has passed, and I guess [this patch](https://github.com/gentoo/libressl/blob/master/dev-lang/python/files/python-3.10.3-libressl.patch) sort of reflects the state of Python v3.10 with LibreSSL v3.5.
It looks like the [next version](https://github.com/openbsd/src/commit/28494c8705b190e170b73f24b26932751c07da84) will support `SSL_CTX_get_security_level` too.
What's left then are those four hash functions, which probably won't be added [anytime](https://github.com/libressl-portable/portable/issues/199) [soon](https://github.com/libressl-portable/portable/issues/455).

Does that change anything for Python w.r.t. LibreSSL support?

**Comment by dhewg** [comment]
Of course I stumble upon [this](https://github.com/python/cpython/blob/main/Doc/whatsnew/3.11.rst#hashlib) just after posting.
That sounds like it can solve the issue about the missing hashing functions?

**Comment by tiran** [comment]
It is unlikely that we will officially support LibreSSL. Even if LibreSSL would become feature compatible with OpenSSL, we just don't have to resources to test and verify that Python works correctly with LibreSSL.

Is LibreSSL still relevant? AFAIK only OpenBSD uses it. All Linux distros have dropped support for LibreSSL, even [Gentoo](https://www.gentoo.org/support/news-items/2021-01-05-libressl-support-discontinued.html). FreeBSD uses OpenSSL and NetBSD seems to prefer OpenSSL as well (they have OpenSSL 1.1.1n and LibreSSL 2.7.4 from 2018).

**Comment by dhewg** [comment]
For the record: Python 3.10.7 builds with LibreSSL 3.7.0 without any further patches.

**Comment by arhadthedev** [comment]
If no deliberate removal of LibreSSL calls is planned, should we keep this PR and its parent issue open?

**Comment by orbea** [comment]
> All Linux distros have dropped support for LibreSSL, even [Gentoo](https://www.gentoo.org/support/news-items/2021-01-05-libressl-support-discontinued.html).

It is still possible to use it on Gentoo with the overlay. https://github.com/gentoo/libressl

**Comment by erlend-aasland** [comment]
@ramikg, can you resolve the conflicts and update the PR?

**Comment by ramikg** [comment]
> @ramikg, can you resolve the conflicts and update the PR?

Sure, done.

**Review by gpshead — CHANGES_REQUESTED** [review]
This PR should not remove `RAND_egd`, that is something OpenSSL provides.

**Comment by ramikg** [comment]
> This PR should not remove `RAND_egd`, that is something OpenSSL provides.

True. I've brought it back, and reworded the comment to be relevant to [OpenSSL](https://github.com/openssl/openssl/blob/5963aa8c196d7c5a940a979299a07418527932af/crypto/rand/rand_egd.c#L20).  
I've kept the `try` clause so that OpenSSL builds with `OPENSSL_NO_EGD` won't break.

(By the way, `RAND_egd` has already been _mostly_ removed from CPython, as I've mentioned [here](https://github.com/python/cpython/pull/25453#issuecomment-1453763157). Also, I didn't expect the Spanish Inquisition.)

**Comment by ramikg** [comment]
I didn't expect the Spanish Inquisition

**Comment by ramikg** [comment]
I have made the requested changes; please review again

**Review by gpshead — APPROVED** [review]

**Pull request merged**